content/blog/2017/05/2017-05-12-kubernetes-journey-on-azure.adoc

```diff
@@ -0,0 +1,397 @@
+
+== The Jenkins way
+
+.Workflow
+----
+  _________________       ____________________       ______________
+  |                |      |                   |      |             |
+  |    Github:     |      |     Jenkins:      |      |   Jenkins:  |
+  | Commit trigger | ---->| Test & Validation | ---->|  Deployment |
+  |________________|      |___________________|      |_____________|
+----
+
+In this approach, Jenkins is used to test, validate and deploy our Kubernetes
+configuration files.  Kubectl can be run on a directory and is idempotent.
+This means that we can run it as often as we want: the result will not change.
+Theoretically, it's the simplest way. The only thing needed is to run
+'kubectl' command each time Jenkins detects changes.
+
+The following Jenkinsfile gives an example of this workflow.
+
+.Jenkinsfile
+[source,groovy]
+----
+  pipeline {
+    agent any
+    stages {
+      stage('Init'){
+        steps {
+          sh 'curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl'
+        }
+      }
+      stage('Test'){
+        steps {
+          sh 'Run tests'
+        }
+      }
+      stage('Deploy'){
+        steps {
+          sh './kubectl apply -R true -f my_project'
+        }
+      }
+    }
+  }
+----
+
+As the devil is in the details, it was not as easy as it looked at first sight.
+We quickly faced several issues.
+
+=== Order matters
+
+Some resources needed to be deployed before others. A workaround was to use
+numbers as file names. But this adds extra logic at file name level, for
+example:
+
+[source]
+----
+project/00-nginx-ingress
+project/09-www.jenkins.io
+----
+
+=== Portability
+
+The deployment environments needed to be the same across development machines
+and Jenkins host. Even if it's a well known challenge, it was not easy to
+achieve.  The more the project grew, the more our scripts needed additional
+tools (`make`, `bats`, `jq` `gpg`, etc).  The more tools we used, the more
+issues appeared because of the different versions used.
+
+Another challenge that emerged when dealing with different environments was:
+how to manage environment's specific configurations (dev, prod, etc).
+
+Is it better to define different configuration files per environment? Perhaps,
+but this means code duplications, or using file templates that require more
+tools (`sed`, `jinja2`, `erb`). And more work, but less code to write and
+maintain.
+
+There isn't a golden rule we discovered, and the answer is probably in between.
+
+In any case, the good news is that `Jenkinsfile` provides an easy way to
+execute tasks from a docker image and an image can contain all the necessary
+tools in our environment, we can even use different docker images for each
+stage along the way.
+
+In the following example, I use the `my_env` Docker image. It contains all the
+tools needed to test, validate, and deploy changes.
```

Comment:
Now here's an Oxford comma.

If you're going to be wrong (i.e. leave it out), at least do it consistently.